tests/integration_toolbox.py

```diff
@@ -156,6 +158,46 @@ def testGetUrlToFileCachedKeepaliveChanged(self):
             conn.close()
         self.assertEqual(b'This is a test\n', data)
 
+    def testGetUrl404(self):
+        """Call get_url for a file that doesn't exist"""
+        with self.assertRaises(HTTPError):
+            data = spacepy.toolbox.get_url(
+                'http://localhost:{}/doesnotexist.txt'.format(self.port))
+    
+    def testGetUrlNoConnNoKeepAlive(self):
+        """Call get_url without specifying connection or keepalive"""
+        ctype = http.client.HTTPConnection
+        connection = ctype("localhost:{}".format(self.port))
+        with self.assertRaises(ValueError):
+            spacepy.toolbox.get_url('http://localhost:{}/'.format(self.port), conn=connection, keepalive=False)
+    
+    def testGetUrlReadableWritable(self):
+        """Call get_url to trigger readable and writable"""
+        ctype = http.client.HTTPConnection
+        connection = ctype("localhost:{}".format(self.port))
+        connection.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM) 
+        _, connection = spacepy.toolbox.get_url('http://localhost:{}/'.format(self.port), conn=connection, keepalive=True)
+        self.assertEqual(connection.sock, None)
+    
+    def testGetUrlValueErrorOn404WithKeepAlive(self):
+        """Call get_url with keepalive and a bad link"""
+        ctype = http.client.HTTPConnection
+        connection = ctype("localhost:{}".format(self.port))
+        with self.assertRaises(RuntimeError):
```

Comment:
Are we actually getting a RuntimeError in the case where a connection is specified (here) but an HTTPError where it isn't (testGetURL404), or is this a case where we're getting lucky because HTTPError is a subclass of RuntimeError? We should capture the most specific exception that we're expecting, and test the message (as mentioned elsewhere).

Reply:
This test was duplicating the functionality of another test, so it was removed